core/src/main/java/org/apache/calcite/rel/rel2sql/CTERelToSqlUtil.java

```diff
@@ -137,34 +175,137 @@ public static void addSqlWithItemNode(SqlWithItem sqlWithItem, List<SqlNode> sql
-      SqlSelect sqlSelect = (SqlSelect) sqlNode;
-      if (sqlSelect.getFrom() instanceof SqlJoin) {
-        SqlNode leftNode = ((SqlJoin) sqlSelect.getFrom()).getLeft();
-        SqlNode rightNode = ((SqlJoin) sqlSelect.getFrom()).getRight();
-
-        //Update Left Node
-        updateNode(leftNode);
-        // Update Right Node
-        updateNode(rightNode);
-
-        //if left and right is join again need recursive call
-      } else if (sqlSelect.getFrom() instanceof SqlBasicCall) {
-        updateNode(sqlSelect.getFrom());
-      } else if (sqlSelect.getFrom() instanceof SqlWithItem) {
-        sqlSelect.setFrom(((SqlWithItem) sqlSelect.getFrom()).name);
+      if (sqlNode instanceof SqlSelect) {
+        SqlSelect sqlSelect = (SqlSelect) sqlNode;
+        // Handle FROM clause
+        SqlNode fromNode = sqlSelect.getFrom();
+        processFromNode(sqlSelect, fromNode);
+        if (isNestedCte(fromNode)
+            &&
+            fromNode instanceof SqlBasicCall
+            &&
+            ((SqlBasicCall) fromNode).getOperator() instanceof SqlAsOperator) {
+          updateNode(fromNode);
+        }
+        // Handle WHERE clause
+        SqlNode whereNode = sqlSelect.getWhere();
+        if (whereNode != null && whereNode instanceof SqlBasicCall) {
+          updateNode(whereNode);
+        }
       }
     }
   }
 
+  private static void processFromNode(SqlNode sqlNode, SqlNode fromNode) {
+    if (fromNode instanceof SqlJoin) {
+      updateSqlJoinNode((SqlJoin) fromNode);
+    } else if (fromNode instanceof SqlBasicCall) {
+      processBasicCall((SqlBasicCall) fromNode);
+    } else if (fromNode instanceof SqlWithItem) {
+      SqlNode query = ((SqlWithItem) fromNode).query;
+
+      if (isNestedCte(query)) {
+        processWithItem((SqlWithItem) fromNode);
+      } else {
+        ((SqlSelect) sqlNode).setFrom(((SqlWithItem) fromNode).name);
+      }
+    }
+  }
+
+  private static void processBasicCall(SqlNode sqlNode) {
+    if (sqlNode instanceof SqlBasicCall) {
+      if (isNestedCte(sqlNode)) {
+        SqlBasicCall basicCall = (SqlBasicCall) sqlNode;
+
+        // Recursively process each operand
+        for (SqlNode operand : basicCall.getOperandList()) {
+          if (operand instanceof SqlSelect) {
+            updateSqlNode(operand);
+          } else if (operand instanceof SqlWithItem) {
+            processWithItem((SqlWithItem) operand);
+          } else if (operand instanceof SqlBasicCall) {
+            processBasicCall(operand); // Recursive call for nested `SqlBasicCall`
+          }
+        }
+      } else {
+        updateNode(sqlNode);
+      }
+    }
+  }
+
+  private static void processWithItem(SqlWithItem withItem) {
+    SqlNode query = withItem.query;
+
+    if (query instanceof SqlSelect) {
+      updateSqlNode(query);
+    } else if (query instanceof SqlBasicCall) {
+      processBasicCall(query);
+    }
+  }
+
+  public static void updateSqlJoinNode(SqlJoin sqlJoin) {
```

Comment:
This method can be simplified like this -
`
public static void updateSqlJoinNode(SqlJoin sqlJoin) {
    updateNodeOrJoin(sqlJoin.getLeft());
    updateNodeOrJoin(sqlJoin.getRight());
}

private static void updateNodeOrJoin(SqlNode node) {
    if (node instanceof SqlJoin) {
        updateSqlJoinNode((SqlJoin) node);  
    } else if (node instanceof SqlBasicCall) {
        updateNode(node); 
    } else {
        updateSqlNode(node);  
    }
}
`